src/hypertable_insert.c

```diff
@@ -106,23 +113,131 @@ static CustomScanMethods hypertable_insert_plan_methods = {
+ * statement.
+ */
 Plan *
-ts_hypertable_insert_plan_create(ModifyTable *mt)
+ts_hypertable_insert_fixup_tlist(Plan *plan)
+{
+	if (IsA(plan, CustomScan))
+	{
+		CustomScan *cscan = (CustomScan *) plan;
+
+		if (cscan->methods == &hypertable_insert_plan_methods)
+		{
+			ModifyTable *mt = linitial(cscan->custom_plans);
+
+			Assert(IsA(mt, ModifyTable));
+
+			cscan->scan.plan.targetlist = copyObject(mt->plan.targetlist);
+			cscan->custom_scan_tlist = NIL;
+		}
+	}
+
+	return plan;
+}
+
+static Plan *
+hypertable_insert_plan_create(PlannerInfo *root,
+							  RelOptInfo *rel,
+							  struct CustomPath *best_path,
+							  List *tlist,
+							  List *clauses,
+							  List *custom_plans)
 {
 	CustomScan *cscan = makeNode(CustomScan);
+	ModifyTable *mt = linitial(custom_plans);
+
+	Assert(IsA(mt, ModifyTable));
 
 	cscan->methods = &hypertable_insert_plan_methods;
 	cscan->custom_plans = list_make1(mt);
-	cscan->scan.plan.lefttree = &mt->plan;
-	cscan->scan.scanrelid = 0;	/* This is not a real relation */
+	cscan->scan.scanrelid = 0;
 
 	/* Copy costs, etc., from the original plan */
 	cscan->scan.plan.startup_cost = mt->plan.startup_cost;
 	cscan->scan.plan.total_cost = mt->plan.total_cost;
 	cscan->scan.plan.plan_rows = mt->plan.plan_rows;
 	cscan->scan.plan.plan_width = mt->plan.plan_width;
-	cscan->scan.plan.targetlist = mt->plan.targetlist;
-	cscan->custom_scan_tlist = NIL;
+
+	/*
+	 * Since this is the top-level plan (above ModifyTable) we need to use the
+	 * same targetlist as ModifyTable. However, that targetlist is not set at
+	 * this point as it is created by setrefs.c at the end of the planning. It
+	 * accounts for things like returning lists that might order attributes in
+	 * a way that does not match the order in the base relation. To get around
+	 * this we use a temporary target list here and later fix it up after the
+	 * standard planner has run.
+	 */
+	cscan->scan.plan.targetlist = copyObject(root->processed_tlist);
+
+	/* Set the custom scan target list for, e.g., explains */
+	cscan->custom_scan_tlist = copyObject(cscan->scan.plan.targetlist);
 
 	return &cscan->scan.plan;
 }
+
+static CustomPathMethods hypertable_insert_path_methods = {
+	.CustomName = "HypertableInsertPath",
+	.PlanCustomPath = hypertable_insert_plan_create,
+};
+
+Path *
+ts_hypertable_insert_path_create(PlannerInfo *root, ModifyTablePath *mtpath)
+{
+	Path	   *path = &mtpath->path;
+	Cache	   *hcache = ts_hypertable_cache_pin();
+	ListCell   *lc_path,
+			   *lc_rel;
+	List	   *subpaths = NIL;
+	Hypertable *ht = NULL;
+	HypertableInsertPath *hipath;
+
+	Assert(list_length(mtpath->subpaths) == list_length(mtpath->resultRelations));
+
+	forboth(lc_path, mtpath->subpaths, lc_rel, mtpath->resultRelations)
+	{
+		Path	   *subpath = lfirst(lc_path);
+		Index		rti = lfirst_int(lc_rel);
+		RangeTblEntry *rte = planner_rt_fetch(rti, root);
+
+		ht = ts_hypertable_cache_get_entry(hcache, rte->relid);
```

Comment:
I think it would be nice to add a check here that we are always talking about the same hypertable here... That we aren't somehow mixing hypertables.